sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala

```diff
@@ -621,20 +621,40 @@ class HDFSFileCatalog(
 
   def getStatus(path: Path): Array[FileStatus] = leafDirToChildrenFiles(path)
 
+  private implicit class LocatedFileStatusIterator(iterator: RemoteIterator[LocatedFileStatus])
+    extends Iterator[LocatedFileStatus] {
+
+    override def hasNext: Boolean = iterator.hasNext
+
+    override def next(): LocatedFileStatus = iterator.next()
+  }
+
   private def listLeafFiles(paths: Seq[Path]): mutable.LinkedHashSet[FileStatus] = {
     if (paths.length >= sqlContext.conf.parallelPartitionDiscoveryThreshold) {
       HadoopFsRelation.listLeafFilesInParallel(paths, hadoopConf, sqlContext.sparkContext)
```

Comment:
Seems we also need to update the `listLeafFiles` that is called by `listLeafFilesInParallel`.